lightning/src/ln/channelmanager.rs

```diff
@@ -6189,6 +6220,255 @@ where
+									session_priv: inter_trampoline_session_priv,
+									previous_hop_data: HTLCPreviousHopData {
+										short_channel_id: prev_short_channel_id,
+										user_channel_id: Some(prev_user_channel_id),
+										counterparty_node_id: prev_counterparty_node_id,
+										channel_id: prev_channel_id,
+										outpoint: prev_funding_outpoint,
+										htlc_id: prev_htlc_id,
+										incoming_packet_shared_secret: incoming_outer_shared_secret,
+										// Phantom payments are only PendingHTLCRouting::Receive.
+										phantom_shared_secret: None,
+										blinded_failure: blinded.map(|b| b.failure),
+										cltv_expiry: Some(incoming_cltv_expiry),
+									},
+									incoming_trampoline_shared_secret,
+									hops: Vec::new(),
+								};
+
+								let mut push_trampoline_forwarding_failure = |msg: String, htlc_source: HTLCSource, forward_scid: Option<u64>, err_code: u16, err_data: Vec<u8>| {
+									let logger = WithContext::from(&self.logger, Some(next_node_id), Some(prev_channel_id), Some(payment_hash));
+									log_info!(logger, "Failed to forward incoming Trampoline HTLC: {}", msg);
+
+									failed_forwards.push((htlc_source, payment_hash,
+										HTLCFailReason::reason(err_code, err_data),
+										HTLCDestination::FailedTrampolineForward { requested_next_node_id: next_node_id, forward_scid }
+									));
+								};
+
+								let next_blinding_point = blinded.and_then(|b| {
+									b.next_blinding_override.or_else(|| {
+										let encrypted_tlvs_ss = self.node_signer.ecdh(
+											Recipient::Node, &b.inbound_blinding_point, None
+										).unwrap().secret_bytes();
+										onion_utils::next_hop_pubkey(
+											&self.secp_ctx, b.inbound_blinding_point, &encrypted_tlvs_ss
+										).ok()
+									})
+								});
+
+								let incoming_amount = match incoming_amt_msat {
+									Some(amount) => amount,
+									None => {
+										push_trampoline_forwarding_failure(format!("Missing incoming amount to calculate routing parameters to next Trampoline hop {next_node_id}"), htlc_source, None, 0x2000 | 25, Vec::new());
+										continue;
+									}
+								};
+
+								let usable_channels: Vec<ChannelDetails> = self.list_usable_channels();
+								let route = match self.router.find_route(
+									&self.node_signer.get_node_id(Recipient::Node).unwrap(),
+									&RouteParameters {
+										payment_params: PaymentParameters {
+											payee: Payee::Clear {
+												node_id: next_node_id,
+												route_hints: vec![],
+												features: None,
+												final_cltv_expiry_delta: 0,
+											},
+											expiry_time: None,
+											max_total_cltv_expiry_delta: incoming_cltv_expiry - outgoing_cltv_value,
+											max_path_count: 1,
+											max_path_length: MAX_PATH_LENGTH_ESTIMATE / 2,
+											max_channel_saturation_power_of_half: 2,
+											previously_failed_channels: vec![],
```

Comment:
Not needed for retries?

Reply:
Yes, that's what I mean. They we need info from previous routing attempts here. But I suppose that'll be added in a follow up that does retries?

I am wondering though if postponing the retries to a later PR might require redoing some of the logic that is added in this one.